File: C/search/Fibonacci_Search.c
// C program implementation for Fibonacci Search 

// Worst case time complexity: Θ(logn)
// Average case time complexity: Θ(log n)
// Best case time complexity: Θ(1)
// Space complexity: Θ(1)


#include <stdio.h>
#include <string.h>
 
int Fibonacci_Search(int a[], int n, long x)
{ 
    int index = 0, position, k;
    static int n_1= -1, n_2 = -1;
    
    // Precomputed Fibonacci numbers from F_0 up to F_46.
    static int fib[]={0, 1, 1, 2, 3, 5, 8, 13, 21, 34, 55, 89, 144, 233, 377, 610, 98,
    1597, 2584, 4181, 6765, 10946, 17711, 28657, 46368, 75025, 121393, 196418, 317811,
    514229, 832040, 1346269, 2178309, 3524578, 5702887, 9227465, 14930352, 24157817,
    39088169, 63245986, 102334155, 165580141};
    if (n_2 != n)
    { 
        k = 0;
        while (fib[k] < n)
            k++;
        n_1 = k;
        n_2 = n;
    }
    else
        k = n_1;
    while (k > 0)
    {
        position = index + fib[--k];
        if ((position >= n) || (x < a[position]));
        else if (x > a[position])
        {
            index = position + 1;
            k--;
        }
        else {
            return position;
        }
    }
    return -1;
}
void main()
{
    int no_element=0;
    printf("Enter the no of element in the array: "); 
        scanf("%d", &no_element);
    int arr[no_element];

    // taking input 
    printf("Enter the array: \n"); 
    for(int z=0; z < no_element; z++){
        scanf("%d", &arr[z]);
    }
    int num;
    printf("Enter an element to search: ");
    scanf("%d", &num);

    (Fibonacci_Search(arr, 10, num) == -1) ? printf("Element NOT found!! ") :
    printf("Element is at index : %d", Fibonacci_Search(arr, 10, num));
 
}
/*
____Example-1____
Enter the no of element in the array: 5                                                                                            
Enter the array:                                                                                                                   
30                                                                                                                                 
20                                                                                                                                 
28                                                                                                                                 
55                                                                                                                                 
62                                                                                                                                 
Enter an element to search: 28                                                                                                     
Element is at index : 2

____Example-2____
Enter the no of element in the array: 6                                                                                            
Enter the array:                                                                                                                   
48                                                                                                                                 
56                                                                                                                                 
24                                                                                                                                 
16                                                                                                                                 
36                                                                                                                                 
66                                                                                                                                 
Enter an element to search: 22                                                                                                     
Element NOT found!!
*/

File: Python/search/fibonacci_search.py
# Fibonacci search implementation in Python3

"""
Purpose:
Fibonacci Search is a comparison-based technique that uses Fibonacci numbers to search an element in a sorted array.

Similarities with binary search:
- worked for sorted arrays
- a divide a conquer algorithm

Return type:
Returns the index of a key in a sorted array or returns -1 if it is not contained
"""

def FibonacciSearch(arr, x, n):
    """
    Inputs-
    arr: A sorted array in which we're going to search for the key
    x: The number to be searched for
    n: Size of the array
    """

    # initialize fibonacci numbers
    Fibm2 = 0  # (m-2)'th Fibonacci No.
    Fibm1 = 1  # (m-1)'th Fibonacci No.
    Fibm = Fibm2 + Fibm1  # Fibm'th Fibonacci

    # Fibm: store the smallest
    # fibonacci Number greater than or equal to n
    while Fibm < n:
        Fibm2 = Fibm1
        Fibm1 = Fibm
        Fibm = Fibm2 + Fibm1

    # marks the eliminated range from front
    offset = -1

    # while there are elements to be inspected.
    # Note: we compare arr[Fibm2] with x.
    # when Fibm becomes 1, Fibm2 becomes 0
    while Fibm > 1:

        # Check if Fibm2 is a valid location
        i = min(offset + Fibm2, n - 1)

        # if x is greater than the value at
        # index Fibm2, cut the subarray array
        # from offset to i
        if arr[i] < x:
            Fibm = Fibm1
            Fibm1 = Fibm2
            Fibm2 = Fibm - Fibm1
            offset = i

        # if x is less than the value at
        # index Fibm2, cut the subarray
        # after i+1
        elif arr[i] > x:
            Fibm = Fibm2
            Fibm1 = Fibm1 - Fibm2
            Fibm2 = Fibm - Fibm1

        # element found, return index
        else:
            return i

    # comparing the last element with x
    if Fibm1 and arr[offset + 1] == x:
        return offset + 1

    # element not found. return -1
    return -1

# taking dynamic input
print("Enter the array in space seperated format: ")
arr = list(map(int, input().split(' ')))
print("Enter the element which is to be searched: ")
x = int(input())
n = len(arr)

# calling function
index = FibonacciSearch(arr, x, n)
if index != -1:
    print("Found at index:", FibonacciSearch(arr, x, n))
else:
    print("Element not found in above listed numbers.")

"""
Sample input:
Enter the array in space seperated format:
10 22 35 40 45 50 80 82 85 90 100
Enter the element which is to be searched:
85

Sample output:
Found at index: 8

Worst case time complexity: O(log n)
Best case time complexity:	O(1)

Space complexity: O(1)
"""
